### PR TITLE
WFCORE-823 Introduce new way for registering transformers

### PR DESCRIPTION
**Pull request by ctomc**
An idea of new way of registering transformers via service loader.

### CODE REVIEW

File: controller/src/main/java/org/jboss/as/controller/extension/ExtensionRegistry.java [hunk 1]
```diff
@@ -532,8 +533,12 @@ public SubsystemRegistration registerSubsystem(String name, ModelVersion version
             if (deprecated){
                 ControllerLogger.DEPRECATED_LOGGER.extensionDeprecated(name);
             }
-            return new SubsystemRegistrationImpl(name, version,
-                    profileRegistration, deploymentsRegistration, extensionRegistryType, extension.extensionModuleName, processType);
+            SubsystemRegistrationImpl result =  new SubsystemRegistrationImpl(name, version,
+                                profileRegistration, deploymentsRegistration, extensionRegistryType, extension.extensionModuleName, processType);
+            if (registerTransformers){
+                transformerRegistry.loadAndRegisterTransformers(name, version, extension.extensionModuleName);
```
Comment (kabir) [inline controller/src/main/java/org/jboss/as/controller/extension/ExtensionRegistry.java]: What about connector which contains several subsystems?

File: controller/src/main/java/org/jboss/as/controller/transform/description/TransformationDescription.java [hunk 2]
```diff
@@ -174,5 +189,27 @@ public static TransformersSubRegistration register(TransformationDescription des
...
+        /**
+         * Register a transformation description as a sub-resource at a given {@linkplain SubsystemRegistration}.
+         *
+         * @param description the subsystem transformation description
+         * @param registration the subsystem registrations
+         * @param range the model version range the transformation applies to
+         * @return the create sub registration
+         */
+        public static TransformersSubRegistration register(TransformationDescription description, SubsystemTransformerRegistration registration, ModelVersionRange range) {
+            final TransformersSubRegistration subRegistration = registration.registerModelTransformers(range, description.getResourceTransformer(),
```
Comment (kabir) [inline controller/src/main/java/org/jboss/as/controller/transform/description/TransformationDescription.java]: Perhaps some wrapper around registration to avoid code duplication with the other register() method?

File: controller/src/main/java/org/jboss/as/controller/extension/ExtensionRegistry.java (hunk 1 above)
Comment (ctomc) [inline controller/src/main/java/org/jboss/as/controller/extension/ExtensionRegistry.java]: hmm, good point. Will look into this.

File: controller/src/main/java/org/jboss/as/controller/transform/description/TransformationDescription.java (hunk 2 above)
Comment (ctomc) [inline controller/src/main/java/org/jboss/as/controller/transform/description/TransformationDescription.java]: Will take a look what could be done.

File: controller/src/main/java/org/jboss/as/controller/extension/ExtensionRegistry.java (hunk 1 above)
Comment (ctomc) [inline controller/src/main/java/org/jboss/as/controller/extension/ExtensionRegistry.java]: Fixes, there is not extra check to register only transformers that are for same subsystem as subsystem we are registering.

File: controller/src/main/java/org/jboss/as/controller/transform/description/TransformationDescription.java (hunk 2 above)
Comment (ctomc) [inline controller/src/main/java/org/jboss/as/controller/transform/description/TransformationDescription.java]: I fixed that, just diff is after this line so it is not seen.

File: controller/src/main/java/org/jboss/as/controller/transform/SubsystemTransformerRegistration.java [hunk 3]
```diff
@@ -0,0 +1,65 @@
+/*
+ * JBoss, Home of Professional Open Source
+ * Copyright 2015, Red Hat, Inc., and individual contributors as indicated
```
Comment (darranl) [inline controller/src/main/java/org/jboss/as/controller/transform/SubsystemTransformerRegistration.java]: Your work is so last year ;-)

**Comment by darranl** [comment]
We may have a little bit of work to do to merge this with our branch assuming it is merged to master but TBH I don't see it as a major issue from a management perspective our changes are just a couple of additional attributes on existing resources to reference new capabilities.

File: controller/src/main/java/org/jboss/as/controller/transform/SubsystemTransformerRegistration.java (hunk 3 above)
Comment (ctomc) [inline controller/src/main/java/org/jboss/as/controller/transform/SubsystemTransformerRegistration.java]: Damn, yeah initial prototype was done than, as I wanted to squeeze it in EAP7 :-)

**Comment by ctomc** [comment]
@bstansberry could we merge this?

File: subsystem-test/framework/src/main/java/org/jboss/as/subsystem/test/AbstractKernelServicesImpl.java [hunk 4]
```diff
@@ -140,7 +140,7 @@ public ModelNode readTransformedModel(ModelVersion modelVersion) {
         return readTransformedModel(modelVersion, true);
     }
 
-    ExtensionRegistry getExtensionRegistry() {
+    protected ExtensionRegistry getExtensionRegistry() {
```
Comment (bstansberry) [inline subsystem-test/framework/src/main/java/org/jboss/as/subsystem/test/AbstractKernelServicesImpl.java]: This method is still only called from within the package.

File: controller/src/main/java/org/jboss/as/controller/extension/SubsystemInformation.java [hunk 5]
```diff
@@ -56,4 +58,25 @@
...
+        return ModelVersion.create(getManagementInterfaceMajorVersion() != null ? getManagementInterfaceMajorVersion() : 0,
+                getManagementInterfaceMinorVersion() != null ? getManagementInterfaceMinorVersion() : 0,
+                getManagementInterfaceMicroVersion() != null ? getManagementInterfaceMicroVersion() : 0
+        );
+    }
+
+
+    default boolean isDeprecated(){
+        return false;
+    }
```
Comment (bstansberry) [inline controller/src/main/java/org/jboss/as/controller/extension/SubsystemInformation.java]: This is an unrelated change and no code is using this method. If we're going to make an unrelated change around this we should drop the getter/setter from SubsystemInformationImpl as no one uses the property.

File: controller/src/main/java/org/jboss/as/controller/extension/SubsystemInformation.java [hunk 6]
```diff
@@ -56,4 +58,25 @@
...
+
+    default boolean isHostCapable() {
+                return false;
+            }
```
Comment (bstansberry) [inline controller/src/main/java/org/jboss/as/controller/extension/SubsystemInformation.java]: This is an unrelated change and no code is using this method. If we're going to make an unrelated change around this we should drop the getter/setter from SubsystemInformationImpl as no one uses the property.

File: controller/src/main/java/org/jboss/as/controller/transform/SubsystemTransformerRegistration.java [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.jboss.as.controller.transform;
+
+import org.jboss.as.controller.ModelVersion;
+import org.jboss.as.controller.ModelVersionRange;
+
+/**
```
Comment (bstansberry) [inline controller/src/main/java/org/jboss/as/controller/transform/SubsystemTransformerRegistration.java]: Needs class javadoc to explain what this interface is.

File: controller/src/main/java/org/jboss/as/controller/transform/TransformerRegistration.java [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+
+/**
+ * @author Tomaz Cerar (c) 2016 Red Hat Inc.
+ */
+public interface TransformerRegistration {
+
+    /** Return name of subsystem this transformer registration is for.
+     * @return non null name of the subsystem
+     */
+    String getSubsystemName();
```
Comment (bstansberry) [inline controller/src/main/java/org/jboss/as/controller/transform/TransformerRegistration.java]: This forces multiple implementations of the ServiceLoader for an extension with multiple subsystems. An alternative is to expose the subsystem name via a getter on SubsystemTransformerRegistration and trust the impl to use that getter and do the right thing.

I don't feel strongly about this though. We only have 1 extension with multiple subsystems anyway. :)

File: controller/src/main/java/org/jboss/as/controller/transform/TransformerRegistration.java [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.jboss.as.controller.transform;
+
+/**
+ * @author Tomaz Cerar (c) 2016 Red Hat Inc.
+ */
+public interface TransformerRegistration {
```
Comment (bstansberry) [inline controller/src/main/java/org/jboss/as/controller/transform/TransformerRegistration.java]: Perhaps ExtensionTransformerRegistration?

The very general name bothers me, as most of our class names in the transformer area are kind of confusing. I can never figure out what is for what. And this interface is clearly limited to use by extensions.

Comment (ctomc) [inline controller/src/main/java/org/jboss/as/controller/transform/TransformerRegistration.java]: That name would definitely be better. Will change it.

File: controller/src/main/java/org/jboss/as/controller/extension/SubsystemInformation.java (hunk 6 above)
Comment (ctomc) [inline controller/src/main/java/org/jboss/as/controller/extension/SubsystemInformation.java]: There was a reason I added that, but I cannot remember what it was. 
If and when I remember again I can send separate PR to do that.

File: controller/src/main/java/org/jboss/as/controller/extension/SubsystemInformation.java (hunk 5 above)
Comment (ctomc) [inline controller/src/main/java/org/jboss/as/controller/extension/SubsystemInformation.java]: This one is tricky, as we have setters as we have "two" phase subsystem initializaton.
to remove setters we would need to set everything in constructor. 
But as getSubsystemInfo() is called by initialize() and initializeParsers() 
problem is that only initialize() has full data needed to construct it without setters.
In most cases parsers are initialized first which create empty SubystemInfoImpl and than when initialize() is called it gets that info and fills it with all extra data.
To properly fix this we would need to do some bigger refactoring which is out of scope of this PR. 
I can look into it afterwards but lets keep it out of this PR for now.

Comment (bstansberry) [inline controller/src/main/java/org/jboss/as/controller/extension/SubsystemInformation.java]: The getter is never used. The field in the implementation is also never read outside the getter. So the setter is pointless.

I don't want a pointless method added to an interface.

File: controller/src/main/java/org/jboss/as/controller/extension/SubsystemInformation.java (hunk 6 above)
Comment (bstansberry) [inline controller/src/main/java/org/jboss/as/controller/extension/SubsystemInformation.java]: After you drop this from this PR, if you remember the reason for it you can restore this method along with the use of it.

File: controller/src/main/java/org/jboss/as/controller/transform/SubsystemTransformerRegistration.java (hunk 7 above)
Comment (ctomc) [inline controller/src/main/java/org/jboss/as/controller/transform/SubsystemTransformerRegistration.java]: added

**Comment by bstansberry** [comment]
This was merged with an added commit on #1752. I don't know why github didn't mark this one as merged too like it normally does.
